From: Karel Zak <kzak@redhat.com>
Date: Fri, 15 Nov 2024 11:32:34 +0100
Subject: meson: add -D tty-setgid=[false|true]

If your distribution does not define permissions for installed
binaries and follows the upstream guidelines, disabling the tty
group's setgid could be a beneficial decision in certain situations.

Signed-off-by: Karel Zak <kzak@redhat.com>
---
 meson.build       | 9 +++++++++
 meson_options.txt | 4 ++++
 2 files changed, 13 insertions(+)

diff --git a/meson.build b/meson.build
index c1f70ab..dc02a21 100644
--- a/meson.build
+++ b/meson.build
@@ -2267,6 +2267,13 @@ if opt
   bashcompletions += ['mesg']
 endif
 
+tty_setgid = get_option('tty-setgid')
+if tty_setgid
+  tty_install_mode = [ 'rwxr-sr-x', 'root', 'tty' ]
+else
+  tty_install_mode = [ false, false, false ]
+endif
+
 opt = not get_option('build-wall').disabled()
 exe = executable(
   'wall',
@@ -2274,6 +2281,7 @@ exe = executable(
   include_directories : includes,
   link_with : [lib_common],
   install_dir : usrbin_exec_dir,
+  install_mode : tty_install_mode,
   install : opt,
   build_by_default : opt)
 if opt
@@ -2292,6 +2300,7 @@ exe = executable(
   include_directories : includes,
   link_with : [lib_common],
   install_dir : usrbin_exec_dir,
+  install_mode : tty_install_mode,
   install : opt,
   build_by_default : opt)
 if opt
diff --git a/meson_options.txt b/meson_options.txt
index 64c9924..050d706 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -201,3 +201,7 @@ option('fs-search-path-extra',
 option('vendordir',
        type: 'string',
        description : 'directory for distribution provided econf files')
+
+option('tty-setgid', type : 'boolean',
+       value : true,
+       description : 'setgid tty group for wall and write programs')
